Refresh documentation

rgreinho opened the pull request:
This is just an idea, so I did not push the research too far, but the
documentation starts to have **A LOT** of information, and the users
have to scroll a lot up and down, so I thought we could think of
refreshing the documentation a little.

As I started to research a new theme for my own projects, I found
[Furo](https://pradyunsg.me/furo/quickstart/) which really looks
awesome!

I think it brings the following benefits:

* It REALLY looks good!
* The side bars (left and right), make the navigation through the
  different sections clearer and quicker.
* We could easily add a "development" section in the left side bar to
  emphasize the contribultion guidelines, changelog, etc.

If we decide to move forward with this idea, there would be some
rewrite/reorganiztion to do, but nothing too crazy.

jdkandersson commented on docs/source/conf.py:
```diff
@@ -23,7 +23,7 @@
 
 # The full version, including alpha/beta/rc tags
-release = "0.2.0"
+release = "2.0.0"
```
I was wondering why that version kept showing up, thanks for fixing!

jdkandersson commented:
I agree, the docs need to be improved. It has been "just add this one more thing here" kind of approach. I also like Furo, it looks much more modern than what is there currently. I'm happy to proceed and help as well!

rgreinho commented:
Awesome! I'll keep working on this branch until the documentation is reorganized to leverage the features of the Furo theme, but without really touching the content and then that will be our starting point to see if we want to change it even further (like splitting sections for instance).

rgreinho commented:
I fixed all the errors related to the embedded tables of content which don't work well with the theme. I also setup `sphinx-autobuild` which makes it easier/faster to iterate while writing the documentation.

rgreinho commented:
As per our conversation, I finished adding the Changelog to the documentation and splitting the `index.rst` file, moving its relevant content to the `advanced.rst` file. 

In addition to that, I also moved the GitHub files (`CONTRIBUTING.md`, `FUNDING.yml`, `CODEOWNERS`) to the `.github` folder for consistency.

Last, I divided the index into 2 main sections, General and Development (see screenshot below) to improve clarity, both for the users and the contributors.

![image](https://user-images.githubusercontent.com/6969134/103177702-4a81e400-4842-11eb-866b-391c39f445c7.png)

jdkandersson commented on docs/source/advanced.rst:
```diff
@@ -0,0 +1,354 @@
+installed. The :samp:`build_yaml` interface accepts the following arguments:
+
+* :samp:`spec_filename`: The name of the OpenAPI specification file. The file
+  must by a YAML file.
```
This wasn't introduced by this change, but I did notice it and it would be great to fix it. This should say `... must be a YAML...` and not `... must by a YAML...`

rgreinho replied:
Addressed ✅

jdkandersson commented on docs/source/advanced.rst:
```diff
@@ -0,0 +1,354 @@
+on a YAML OpenAPI specification which has been extended with any relevant
+OpenAlchemy extension properties.
+
+The :samp:`build_yaml` interface requires the :samp:`PyYAML` library to be
```
Again not specifically related to this PR but something I noticed, do you think there should be a line here that shows how to install with yaml? I.e. `pip install OpenAlchemy[yaml]` I believe

rgreinho replied:
I can add it here if you really want to, but we already mention that in the `README.rst` file (in the installation section), so I think the information would be redundant.

jdkandersson replied:
Fair point :smile:

jdkandersson commented:
Do you think it is worth including what you found at the company you work at regarding pylint as a section in the `advanced.rst` file? I.e. to add a statement to ignore? I can't quite remember what the line was

jdkandersson commented on docs/source/advanced.rst:
```diff
@@ -0,0 +1,354 @@
+* The object and property descriptions from the OpenAPI specification in the
+  class and function docstrings.
+
+.. _backrefs:
```
This section on `backrefs` feels out of place here, should remove it to just before `.. _model-utilities:`?

rgreinho replied:
Not sure about that. to me it is not really a model utility either. It is more related to the extension properties I think. So maybe move it to an example instead?

jdkandersson replied:
Reading it again, I think it is ok. It is a side note that explains a piece of internal functionality so might be misplaced in examples. Let's leave it here for now :smile:

jdkandersson commented on docs/source/advanced.rst:
```diff
@@ -0,0 +1,354 @@
+
+* The dictionary based on which the model is constructed is checked against
+  the schema used to define the model.
+* If the model includes a relationship, the relationship is constructed
```
btw, this is another advantage of `from_dict`, the case where destructing a dictionary into a model constructor does not work is in the case of a relationship, where the relationship parameter needs to be the model and not the dictionary representation of it, but it is possible to destruct a dictionary into `from_dict` in that case as well

jdkandersson commented on docs/source/index.rst:
```diff
@@ -91,69 +89,20 @@ The :samp:`init_json` interface is similar to the :ref:`init-yaml` interface
 
 Used to build a package with the SQLAlchemy models (including type hints) based
 on a YAML OpenAPI specification which has been extended with any relevant
 OpenAlchemy extension properties.
```
It would be good to include a link to read more here, also the case for the other sections where there is a short introduction

jdkandersson commented on pyproject.toml:
```diff
@@ -68,6 +68,12 @@ PyYAML = "^5.3.1"
 typeguard = "^2.10.0"
 sqlalchemy_mixins = "^1.2.1"
 wheel = "^0.36.2"
+pip = "^20.3.3"
```
are `pip` and `install` intentional?

rgreinho replied:
Oops, my mistake. I'll remove that.

jdkandersson commented:
This is excellent, thank you for contributing it. There are a few minor comments then it is good to go!

Is moving to GitHub docs something we should do in another issue?

jdkandersson commented:
Should the depandabot file go to the workflows folder? I tried that in another repository and it threw an error, maybe I was doing it wrong there?

rgreinho commented:
> Should the depandabot file go to the workflows folder? I tried that in another repository and it threw an error, maybe I was doing it wrong there?

Ah, no the dependabot file is not a workflow, just a configuration file. If you want I can add a workflow to automatically merge PRs coming from dependabot.

rgreinho commented:
> Do you think it is worth including what you found at the company you work at regarding pylint as a section in the `advanced.rst` file? I.e. to add a statement to ignore? I can't quite remember what the line was

Ah yes! I added it ✅

jdkandersson commented:
> > Should the depandabot file go to the workflows folder? I tried that in another repository and it threw an error, maybe I was doing it wrong there?
> 
> Ah, no the dependabot file is not a workflow, just a configuration file. If you want I can add a workflow to automatically merge PRs coming from dependabot.

I would prefer not to have a workflow that auto merges the depandabot PRs for now because I don't find myself always just merging the changes. In some cases the changes to the lock file are wrong or (mostly in Node I have found so far) and in other cases I prefer a single commit. As they improve depandabot and I find myself just merging the PRs then I think it would be good to automate, at least those are my thoughts at the moment.

jdkandersson commented:
> > Do you think it is worth including what you found at the company you work at regarding pylint as a section in the `advanced.rst` file? I.e. to add a statement to ignore? I can't quite remember what the line was
> 
> Ah yes! I added it ✅

Maybe I didn't notice, I think it was added as a line in `.pylintrc`? I think it is a good thing to have, I also meant to put a paragraph in the documentation (e.g. in the `advanced.rst` file) maybe on `pylint` that says something like:

## Pylint

The auto-generated models file don't play nice with the pylint rule `no-member` (E1101) because the models are generated in code which is not currently supported by pylint. To avoid spurious pylint errors add the following line to your `.pylintrc` file in the `[IMPORTS]` section:

```generated-members=models.*```

rgreinho commented:
> > > Do you think it is worth including what you found at the company you work at regarding pylint as a section in the `advanced.rst` file? I.e. to add a statement to ignore? I can't quite remember what the line was
> > 
> > 
> > Ah yes! I added it ✅
> 
> Maybe I didn't notice, I think it was added as a line in `.pylintrc`? I think it is a good thing to have, I also meant to put a paragraph in the documentation (e.g. in the `advanced.rst` file) maybe on `pylint` that says something like:
> ## Pylint
> 
> The auto-generated models file don't play nice with the pylint rule `no-member` (E1101) because the models are generated in code which is not currently supported by pylint. To avoid spurious pylint errors add the following line to your `.pylintrc` file in the `[IMPORTS]` section:
>
> `generated-members=models.*`

Ah no, I meant "I added it after reading your comment" 😁 

I'll add the paragraph too, that makes sense 👍

rgreinho commented:
All remarks should have been addressed ✅

jdkandersson commented:
Excellent, thank you, merged 🎉
